crates/cext/src/circuit.rs

```diff
@@ -0,0 +1,483 @@
+        unsafe {
+            let _ = Box::from_raw(circuit);
+        }
+    }
+}
+
+/// @ingroup QkCircuit
+/// Append a standard gate to the circuit.
+///
+/// @param circuit A pointer to the circuit to add the gate to.
+/// @param gate The StandardGate to add to the circuit.
+/// @param qubits The pointer to the array of ``uint32_t`` qubit indices to add the gate on.
+/// @param params The pointer to the array of ``double`` values to use for the gate parameters.
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///
+/// # Safety
+///
+/// The ``qubits`` and ``params`` types are expected to be a non-null pointer to an array of
+/// ``uint32_t`` and ``double`` respectively where the length is matching the expectations for
+/// the standard gate. If the array is insufficently long the behavior of this function is
+/// undefined as this will read outside the bounds of the array.
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_append_standard_gate(
+    circuit: *mut CircuitData,
+    gate: StandardGate,
+    qubits: *const u32,
+    params: *const f64,
+) -> ExitCode {
+    let circuit = unsafe { mut_ptr_as_ref(circuit) };
+    unsafe {
+        let qargs: &[Qubit] = match gate.num_qubits() {
+            0 => &[],
+            1 => &[Qubit(*qubits.wrapping_add(0))],
+            2 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+            ],
+            3 => &[
+                Qubit(*qubits.wrapping_add(0)),
+                Qubit(*qubits.wrapping_add(1)),
+                Qubit(*qubits.wrapping_add(2)),
+            ],
+            // There are no standard gates > 3 qubits
+            _ => unreachable!(),
+        };
+        let params: &[Param] = match gate.num_params() {
+            0 => &[],
+            1 => &[(*params.wrapping_add(0)).into()],
+            2 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+            ],
+            3 => &[
+                (*params.wrapping_add(0)).into(),
+                (*params.wrapping_add(1)).into(),
+                (*params.wrapping_add(2)).into(),
+            ],
+            // There are no standard gates that take > 3 params
+            _ => unreachable!(),
+        };
+        circuit.push_standard_gate(gate, params, qargs);
+    }
+    ExitCode::Success
+}
+
+/// @ingroup QkCircuit
+/// Append a measurement to the circuit
+///
+/// @param circuit A pointer to the circuit to add the gate to
+/// @param qubits The ``uint32_t`` for the qubit to measure
+/// @param clbits The ``uint32_t`` for the clbit to store the measurement outcome in
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_measure(qc, 0, 0);
```

Comment:
The circuit doesn't have any classical bits, do we need to add one?
```suggestion
///     QkCircuit *qc = qk_circuit_new(100, 1);
///     qk_circuit_append_measure(qc, 0, 0);
```